agents/gmail/main.py

```diff
@@ -0,0 +1,100 @@
+import json
+import random
+from typing import Optional
+from threading import Thread
+
+import fixieai
+from llama_index import GPTSimpleVectorIndex, download_loader
+
+import os
+
+BASE_PROMPT = """I'm an agent that answers questions about Gmail messages."""
+FEW_SHOTS = """
+Q: What are the next steps with Oana from SignalFire?
+Ask Func[question]: What are the next steps with Oana from SignalFire?
+Func[question] says: The next steps with Oana from SignalFire are to schedule a meeting for the week of March 20th.
+A: The next steps with Oana from SignalFire are to schedule a meeting for the week of March 20th.
+
+Q: What is Nick's start date?
+Ask Func[question]: What is Nick's start date?
+Func[question] says: Nick Heiner's start date is April 4.
+A: Nick Heiner's start date is April 4.
+"""
+
+# Configure OAuth params.
+with open("credentials.json") as f:
+    credentials = json.load(f)["installed"]
+oauth_params = fixieai.OAuthParams(
+    client_id=credentials["client_id"],
+    auth_uri=credentials["auth_uri"],
+    token_uri=credentials["token_uri"],
+    client_secret=credentials["client_secret"],
+    scopes=["https://www.googleapis.com/auth/gmail.readonly"],
+)
+
+# Initialize agent.
+agent = fixieai.CodeShotAgent(BASE_PROMPT, FEW_SHOTS, oauth_params=oauth_params)
+
+initializing = False
+ready = False
+status = "Initializing"
+index: Optional[GPTSimpleVectorIndex] = None
+
+
+def init() -> str:
```

Comment:
this is especially complicated because you need to store these vectors on a per-user basis, rather than a one-time init. So async agent init won't really help you here. I would download the loader at agent startup (I assume that is reasonably fast).

The indexing process, I would treat that like OAuth. If, after OAuth, you get a query and you don't have a `vectorstore` key in user storage, say that you're indexing and kick off the process. (We should probably have a SDK hook for when OAuth completes that we could use to jump-start this). If the read is fast enough, you could just block here waiting for the index to complete. (This is another place where returning async status messages would be useful).

Reply:
Some interesting platform ideas here to make this super simple for devs:
- just store the plaintext documents in userstorage, and we'll embed them (no need to embed them/call openai yourself)
- via a new `fixie_query_storage` func to run a LLM query against a set of stored docs (similar to `fixie_query_corpus` and `fixie_query_embed`)